### PR TITLE
gnrc_sock_udp: accept response from any address if remote is multicast

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

`sock_udp_create()` with a set remote is not just a convenient way to not having to keep the remote around when calling `sock_udp_send()` it actually binds the socket to that remote and will drop any response from a different address.

This is of course not very practical when the remote is a multicast address as this means all unicast responses are dropped.

This adds a new `SOCK_FLAGS_BIND_REMOTE` flag that gets set automatically for unicast remotes to retain current behavior.

This also moves the remote-checking code `sock_udp_recv_buf_aux()` to make it more readable as this was not so straightforward to debug. 

### Testing procedure

A new test case was added to `tests/gnrc_sock_udp`.

### CODE REVIEW

File: sys/include/net/sock.h [hunk 1]
```diff
@@ -144,6 +144,7 @@ extern "C" {
  * @{
  */
 #define SOCK_FLAGS_REUSE_EP     (0x0001)    /**< allow to reuse end point on bind */
+#define SOCK_FLAGS_BIND_REMOTE  (0x0002)    /**< restrict responses to remote address */
```
Comment (miri64) [inline sys/include/net/sock.h:147]: Isn't the "binding" of a remote typically a `connect`? Given that there are relations between `sock` and the POSIX socket API, using the wording of "bind" might be confusing to the reader here.

```suggestion
#define SOCK_FLAGS_CONNECT_REMOTE   (0x0002)    /**< restrict responses to remote address */

Comment (benpicco) [reply sys/include/net/sock.h]: fine with me

**Review by kfessel — APPROVED** [review]
read this looks good, and i trust @benpicos testing

**Comment by benpicco** [comment]
Thank you for the review!

**Pull request merged**